webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/servers/RemoveNodeService.scala

```diff
@@ -327,69 +493,192 @@ class RemoveNodeServiceImpl(
+    NodeLoggerPure.Delete.debug(s"  - clean-up node '${nodeId.value}' policy files in /var/rudder/share") *>
+    cleanPoliciesRec(nodeId, File(varRudderShare)).runDrain.catchAll(err =>
+      NodeLoggerPure.Delete.error(s"Error when cleaning policy files for node ${(nodeId, info).name}: ${err.fullMsg}")
+    )
   }
 
+  /*
+   * We want to delete the policy directory for node.
+   * We want to look in all policy servers, b/c rudder used to let files everywhere.
+   * To be more resilient, we only look at directory structure, not what rudder believes.
+   * Structure looks like that:
+   * /var/rudder/share
+   *   |- 72F5478E-8214-4B90-8923-D1B1D944DFF5  // simple node
+   *   |    `- rules
+   *   |- a271f3af-0f2b-4538-a1da-0f7b1b1900f2  // relay
+   *        |- rules // relay own policies
+   *        `- share // relay nodes
+   *             |- ca549eb1-6b4b-4823-b7ec-fc5a818893d0 // simple node
+   *             |    `- rules
+   *             `- b8dcd41d-be6f-40e5-9b4f-f494fdeda113 // sub-relay
+   *                  |- rules
+   *                  `- share
+   *                       |- 93e375c4-914e-49b0-98d2-c2aa3b3a6095
+   *                       ... etc ..
+   *
+   * So, algo structure is recursive by nature, and we do
+   * - start at /var/rudder/share [policy server root directory]
+   * - look for node ID directory, delete it
+   * - select all relays by looking for /share, recurse.
+   *
+   * That means that we will have to walk all node directory at least once.
+   * This can be costly when there is 20 000 nodes but:
+   * - we can do that in an iterator and only consume one fd at a time
+   * - it only happens when a node is deleted, so that should not even been in the
+   *   "several time per minute" range.
+   */
+  def cleanPoliciesRec(nodeId: NodeId, file: File): ZStream[Any, RudderError, File] = {
+    ZStream.fromEffect(
+      IOResult.effect(file.exists).catchAll(err =>
+        NodeLoggerPure.Delete.error(err.fullMsg) *> false.succeed
+      )
+    ).flatMap(cond =>
+      // here file is either root or a rec call with "path/share". If it doesn't exists,
+      // it's because the node wasn't a relay, so processing for that branch.
+      if(!cond) ZStream.empty
+      else {
+        ZStream.fromIterator(IOResult.effect(file.children)).tap(_ =>
+            NodeLoggerPure.Delete.ifTraceEnabled {
+              (ZIO.effect {
+                (root / "proc" / "self").children.size
+              } catchAll { _ => (-1).succeed }).flatMap(openfd =>
+                NodeLoggerPure.Delete.trace(s"Currently ${openfd} open files by Rudder Server process")
+              )
+            }
+        ).flatMap { nodeFolder =>
+          if(nodeId == NodeId(nodeFolder.name)) {
+            ZStream.fromEffect(IOResult.effect(nodeFolder.delete()) *> nodeFolder.succeed)
+          } else {
+            cleanPoliciesRec(nodeId, nodeFolder / "share")
```

Comment:
are you sure it doesn't use too much file descriptor?

Reply:
Looks a lot like https://github.com/Normation/rudder/pull/2591/files#diff-da04ceb9ece0de028bbd21114d6f6b8ebcff2ecce0791be224dd27e3456041cfR60 i need to look if it differs much or not (we limit to a node id and that already something different)

Reply:
Yes it use 1 file descriptor, you can see the count with logger `nodes.delete`